### PR TITLE
fix bug of Qwen3 moe

### PR DESCRIPTION
**Pull request by n1ck-guo**

### CODE REVIEW

File: auto_round/data_type/gguf.py [hunk 1]
```diff
@@ -405,28 +425,36 @@ def iterative_wls_quant_search(data, bits=4, rrmin=-1.0, rdelta=0.1, nstep=20, u
...
+        this_min[this_min > 0] = 0
+        this_scale[this_min > 0] = (sum_xl / sum_l2)[this_min > 0]
 
-        quant_data = torch.clamp(torch.round((1 / this_scale) * (data - this_min)), minq, maxq)
-        diff = this_scale * quant_data + this_min - data
-        mad = torch.sum(weights * (torch.abs(diff) if use_mad else diff ** 2), dim=1, keepdim=True)
+        # quant_data = torch.clamp(torch.round((1 / this_scale) * (data - this_min)), minq, maxq)
+        # diff = this_scale * quant_data + this_min - data
+        diff = this_scale * quant_data_new + this_min - data
+        mad = torch.sum(weights * (torch.abs(diff) if use_mad else diff ** 2), dim=-1, keepdim=True)
```
Comment (wenhuach21) [inline auto_round/data_type/gguf.py]: weights*diff**2

File: auto_round/data_type/gguf.py [hunk 2]
```diff
@@ -405,28 +425,36 @@ def iterative_wls_quant_search(data, bits=4, rrmin=-1.0, rdelta=0.1, nstep=20, u
...
-    scale = 1 / ((maxq - minq) / (rmax - rmin + 1e-8))
-    quant_data = torch.clamp(torch.round((maxq - minq) / (rmax - rmin + 1e-8) * (data - rmin)), minq, maxq)
+    # scale = 1 / ((maxq - minq) / (rmax - rmin + 1e-8))
+    scale = (rmax - rmin) / (maxq - minq)
+    iscale = torch.where(scale == 0, 0, 1 / scale)
+    # quant_data = torch.clamp(torch.round((maxq - minq) / (rmax - rmin + 1e-8) * (data - rmin)), minq, maxq)
+    quant_data = torch.clamp(torch.round(iscale * (data - rmin)), minq, maxq)
     diff = scale * quant_data + rmin - data
-    best_mad = torch.sum(weights * (torch.abs(diff) if use_mad else diff ** 2), dim=1, keepdim=True)
+    best_mad = torch.sum(weights * (torch.abs(diff) if use_mad else diff ** 2), dim=-1, keepdim=True)
```
Comment (wenhuach21) [inline auto_round/data_type/gguf.py]: same

File: auto_round/export/export_to_gguf/convert.py [hunk 3]
```diff
@@ -1130,6 +1129,30 @@ def __init_subclass__(cls):
         # would require using decorated functions instead of simply defining the property
         if "model_arch" not in cls.__dict__:
             raise TypeError(f"Missing property 'model_arch' for {cls.__name__!r}")
+    
+    def get_moe_name(self, name, new_name):
+        type_mapping = {
+            "FFN_GATE_EXP": ["gate_proj", "w1", "linear"],
+            "FFN_DOWN_EXP": ["down_proj", "w2", "linear_1"],
```
Comment (wenhuach21) [inline auto_round/export/export_to_gguf/convert.py:1136]: is this code general?

File: auto_round/data_type/gguf.py [hunk 4]
```diff
@@ -137,9 +137,10 @@ def double_quant_tensor(tensor, bits):
     tensor = tensor.to(torch.float32)  # Ensure tensor is in float32 for precision
     maxq = 2 ** bits - 1
     wmax = torch.clamp(tensor.max(-1)[0], min=0)
-    scale = wmax / maxq
+    scale = torch.where(wmax > 0, wmax / maxq, 0)
```
Comment (wenhuach21) [inline auto_round/data_type/gguf.py]: just `scale/maxq` is enough

File: auto_round/data_type/gguf.py [hunk 5]
```diff
@@ -411,29 +414,37 @@ def iterative_wls_quant_search(data, bits=4, rrmin=-1.0, rdelta=0.1, nstep=20, u
...
         D = sum_w * sum_l2 - sum_l ** 2
         this_scale = (sum_w * sum_xl - sum_x * sum_l) / D
         this_min = (sum_l2 * sum_x - sum_l * sum_xl) / D
+        this_min[this_min > 0] = 0
+        this_scale[this_min > 0] = (sum_xl / sum_l2)[this_min > 0]
 
-        quant_data = torch.clamp(torch.round((1 / this_scale) * (data - this_min)), minq, maxq)
-        diff = this_scale * quant_data + this_min - data
-        mad = torch.sum((weights * torch.abs(diff)) if use_mad else weights * diff ** 2, dim=1, keepdim=True)
+        # quant_data = torch.clamp(torch.round((1 / this_scale) * (data - this_min)), minq, maxq)
```
Comment (wenhuach21) [inline auto_round/data_type/gguf.py]: Revert this change, as it improves accuracy.

File: auto_round/data_type/gguf.py [hunk 6]
```diff
@@ -304,6 +305,7 @@ def quant_tensor_gguf_asym_dq(
         Tuple: (Quantized-dequantized tensor, scale dictionary, zero-point dictionary)
     """
     orig_dtype = tensor.dtype
+    tensor = tensor.to(torch.float32)
```
Comment (wenhuach21) [inline auto_round/data_type/gguf.py]: delete this, same with L310

File: auto_round/data_type/gguf.py [hunk 7]
```diff
@@ -345,13 +346,13 @@ def quant_tensor_gguf_asym_dq(
...
-        # if imatrix is None:
-        #     av_x = torch.sqrt(sigma2)
-        #     quant_weights = torch.abs(av_x + tensor * tensor)
-        # else:
-        #     imatrix = imatrix.reshape(1, -1).expand(tensor.numel() // imatrix.numel(), -1).reshape(tensor.shape)
-        #     quant_weights = imatrix * torch.sqrt(sigma2 + tensor * tensor)
+        sigma2 = torch.sum(tensor ** 2, dim=-1, keepdim=True) / QK_K
+        if imatrix is None:
+            av_x = torch.sqrt(sigma2)
+            quant_weights = torch.abs(av_x + tensor * tensor)
```
Comment (wenhuach21) [inline auto_round/data_type/gguf.py]: comment these codes as we only use imatrix and don't use other tricky weights

File: auto_round/data_type/gguf.py [hunk 8]
```diff
@@ -411,29 +413,37 @@ def iterative_wls_quant_search(data, bits=4, rrmin=-1.0, rdelta=0.1, nstep=20, u
...
+        sum_l2 = torch.sum(mul_weights_quant_data * quant_data_new, dim=-1, keepdim=True)
+        sum_xl = torch.sum(mul_weights_quant_data * data, dim=-1, keepdim=True)
 
         D = sum_w * sum_l2 - sum_l ** 2
         this_scale = (sum_w * sum_xl - sum_x * sum_l) / D
         this_min = (sum_l2 * sum_x - sum_l * sum_xl) / D
+        this_min[this_min > 0] = 0
+        this_scale[this_min > 0] = (sum_xl / sum_l2)[this_min > 0]
 
         quant_data = torch.clamp(torch.round((1 / this_scale) * (data - this_min)), minq, maxq)
```
Comment (wenhuach21) [inline auto_round/data_type/gguf.py]: maybe we should use reverse_this_scale

**Review by wenhuach21 — APPROVED** [review]

File: test/test_cpu/test_gguf_format.py [hunk 9]
```diff
@@ -237,6 +237,35 @@ def test_gguf_baseline(self):
...
+            disable_opt_rtn=True 
+        )
+        quantized_model_path = "./saved"
+        autoround.quantize_and_save(output_dir=quantized_model_path, format="gguf:q4_k_m")
+        self.assertEqual(autoround.layer_config["model.layers.11.self_attn.v_proj"]["super_group_size"], 16)
+        self.assertEqual(autoround.layer_config["model.layers.11.self_attn.v_proj"]["data_type"], "int_sym_dq")
+        self.assertEqual(autoround.layer_config["model.layers.7.self_attn.v_proj"]["data_type"], "int_asym_dq")
+        self.assertEqual(autoround.model.model.layers[0].self_attn.v_proj.bits, 6)
+        self.assertEqual(autoround.model.model.layers[12].self_attn.v_proj.bits, 4)
+        shutil.rmtree("./saved", ignore_errors=True)
```
Comment (wenhuach21) [inline test/test_cpu/test_gguf_format.py:271]: add lm_head and embedding layer bits checker

File: test/test_cpu/test_gguf_format.py [hunk 10]
```diff
@@ -237,6 +237,35 @@ def test_gguf_baseline(self):
...
+        autoround = AutoRound(
+            model,
+            tokenizer,
+            bits=bits,
+            group_size=group_size,
+            sym=sym,
+            iters=0,
+            seqlen=1,
+            dataset=self.llm_dataloader,
+            data_type="int_asym_dq",
```
Comment (wenhuach21) [inline test/test_cpu/test_gguf_format.py]: remove all the quantization config setting as we have set the format

File: test/test_cpu/test_gguf_format.py [hunk 11]
```diff
@@ -237,6 +237,35 @@ def test_gguf_baseline(self):
...
+        model_name = "Qwen/Qwen2.5-7B-Instruct"
+        model_name = "/models/Qwen2.5-7B-Instruct"
+        model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype="auto", trust_remote_code=True)
+        tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
+        bits, group_size, sym, super_group_size, super_bits = 4, 32, False, 8, 6
+        autoround = AutoRound(
+            model,
+            tokenizer,
+            bits=bits,
+            group_size=group_size,
```
Comment (wenhuach21) [inline test/test_cpu/test_gguf_format.py]: add a layer_config to set embedding,lm_head, and some other layers and then check the bits
